### PR TITLE
feat: offline token validation

### PR DESCRIPTION
**Pull request by florind12**
Added offline token validation method.

## Description

This PR requires changes in `aio-lib-ims`: https://github.com/adobe/aio-lib-ims/pull/92

## Related Issue
https://github.com/adobe/aio-cli-plugin-app/issues/386

## How Has This Been Tested?
Unit tests.

Manual QA. 
Steps: 
1. Login via `aio auth`.
2. Modify `access_token`. 
3. `aio app init`

Expected: 
- Refresh token is used to generate new `access_token`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/helpers.js [hunk 1]
```diff
@@ -61,6 +65,64 @@ function readFileString (file) {
...
+ * @param {string} env IMS environment type
+ * @param {string} certName certificate that can be used to verify the signature
+ * @returns {Buffer | Promise<Buffer>} cert
+ */
+async function _readOrGetCert (env, certName) {
+  try {
+    return fs.readFileSync(path.join(CONFIG_PATH, certName))
+  } catch (e) {
+    if (env && certName) {
+      const URL = `https://static.adobelogin.com/keys/${env}/${certName}`
```
Comment (shazron) [inline src/helpers.js:80]: let's use https://github.com/adobe/aio-lib-core-networking with exponential backoff and use async/await.

**Review by shazron — CHANGES_REQUESTED** [review]

File: src/helpers.js [hunk 2]
```diff
@@ -61,6 +65,64 @@ function readFileString (file) {
...
+ * Read a cached cert when available or get a new one.
+ * @private
+ * @param {string} env IMS environment type
+ * @param {string} certName certificate that can be used to verify the signature
+ * @returns {Buffer | Promise<Buffer>} cert
+ */
+async function _readOrGetCert (env, certName) {
+  try {
+    return fs.readFileSync(path.join(CONFIG_PATH, certName))
+  } catch (e) {
```
[the remark this answers is not in the document]
Comment (florind12) [reply src/helpers.js:78]: I totally forgot I can do that.

File: src/helpers.js [hunk 3]
```diff
@@ -61,6 +65,64 @@ function readFileString (file) {
...
+    if (env && certName) {
+      const URL = `https://static.adobelogin.com/keys/${env}/${certName}`
+      return new Promise((resolve, reject) => {
+        aioLogger.debug('No cached certificate found, grab new one.')
```
Comment (shazron) [inline src/helpers.js:82]: "No cached certificate found, grabbing a new one." Can we also log the url, and where we are saving the cert

**Review by shazron — CHANGES_REQUESTED** [review]

File: src/helpers.js [hunk 4]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
   })
 }
 
+/**
+ * Load local private key
```
Comment (shazron) [inline src/helpers.js:71]: this is a public key. We have to be careful with naming, and for security review. Private keys are never ever downloaded

File: src/helpers.js [hunk 5]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
...
+ * @private
+ * @param certName name of the private key
+ * @returns {Buffer| undefined} privateKey
+ */
+function loadPrivateKey (certName) {
```
Comment (shazron) [inline src/helpers.js:76]: loadPublicKey

File: src/helpers.js [hunk 6]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
   })
 }
 
+/**
+ * Load local private key
+ * @private
+ * @param certName name of the private key
```
Comment (shazron) [inline src/helpers.js:73]: public key

File: src/helpers.js [hunk 7]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
...
+ * @returns {Buffer| undefined} privateKey
+ */
+function loadPrivateKey (certName) {
+  if (certName) {
+    const certLocation = path.join(CONFIG_PATH, certName)
+    if (fs.existsSync(certLocation)) {
+      aioLogger.debug('Private key loaded from: %s', certLocation)
```
Comment (shazron) [inline src/helpers.js:80]: public key

File: src/helpers.js [hunk 8]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
...
+  }
+}
+/**
+ * Fetch new private key.
+ * @private
+ * @param {string} env IMS environment type
+ * @param {string} certName Certificate name that can be used to verify the signature
+ * @returns {Promise<string>} privateKey
+ */
+async function getPrivateKey (env, certName) {
```
Comment (shazron) [inline src/helpers.js:92]: getPublicKey

File: src/helpers.js [hunk 9]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
...
+  if (env && certName) {
+    const URL = `https://static.adobelogin.com/keys/${env}/${certName}`
+    const certLocation = path.join(CONFIG_PATH, certName)
+    aioLogger.debug('Fetching a new private key...')
```
Comment (shazron) [inline src/helpers.js:96]: public key

File: src/helpers.js [hunk 10]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
...
+    const fetchRetry = new HttpExponentialBackoff()
+    const res = await fetchRetry.exponentialBackoff(URL, {})
+    const data = await res.text()
+    /* istanbul ignore else */
+    if (isPrivateKey(data)) {
+      fs.writeFileSync(certLocation, data)
+      aioLogger.debug('Private key saved at: %s', certLocation)
```
Comment (shazron) [inline src/helpers.js:103]: public key

File: src/helpers.js [hunk 11]
```diff
@@ -62,6 +67,70 @@ function readFileString (file) {
...
+  return Promise.reject(new Error('Invalid params.'))
+}
+
+/**
+ * Validates provided token using an IMS public key cert.
+ *
+ * @param {string} token Jwt token you want to validate. It validates only the access_token type.
+ * @returns {Promise<void>} Will throw if token is invalid
+ */
+async function verifyJwt (token) {
```
Comment (shazron) [inline src/helpers.js:117]: The whole implementation is incorrect. This should not use jwt.verify (call it offlineTokenValidate or something to that effect), but verify the token according to the 6 clauses in the "Validating a token" section of the internal wiki page.

**Review by shazron — CHANGES_REQUESTED** [review]

**Pull request closed without merging**